### PR TITLE
feat: Limited login and Profile for IOS with type def

### PR DESCRIPTION
**Pull request by cocoa-rum**
PR for https://github.com/thebergamo/react-native-fbsdk-next/issues/7

Thanks  [@leolusoli](https://github.com/leolusoli)  for **_Profile_** part

PR contains Authentication Token along side with limited log in:
If we use limited login, we lose access to access token, so we need an alternative and Authentication Token is. So from my view point Authentication Token complements limited log in

Also Profile part
Related to the [commnent](https://github.com/thebergamo/react-native-fbsdk-next/issues/7#issuecomment-819483813)

### CODE REVIEW

File: src/FBAuthenticationToken.js
Comment (bernard-kms) [inline src/FBAuthenticationToken.js:50]: I understand AuthenticationToken only valid in iOS, but I think we can give more information to the developers about this platform-specific function via name or warnings. Something like `-IOS` suffix perhaps? e.g. `getAuthenticationTokenIOS`

File: src/FBProfile.js
Comment (bernard-kms) [inline src/FBProfile.js:86]: I think this could be somewhat unexpected as Profile functionality is also implemented in Android SDK: https://developers.facebook.com/docs/reference/androidsdk/current/facebook/com/facebook/profile.html

I think it would be nice to implement Android side as well.

**Review by bernard-kms — COMMENTED** [review]
I just leave some thoughts about platform-specific behavior. I think it would be more safe with careful implementation because not everyone well understand about each platform's limitations, especially in cross-platform libraries like react native.

**Comment by cocoa-rum** [comment]
@bernard-kms I agree. I will improve what you said
Speaking about **_"platform-specific behavior"_**. Log in button contains `nonce` and `loginTracking`, which are specifice for IOS too, should I rename them `nonceIOS` and `loginTrackingIOS` as well ?

**Comment by bernard-kms** [comment]
> Speaking about **_"platform-specific behavior"_**. Log in button contains `nonce` and `loginTracking`, which are specifice for IOS too, should I rename them `nonceIOS` and `loginTrackingIOS` as well ?

I think that would be safer(maybe the safest) way to implement it. When limited login available in android (FB would implement it, wouldn't they?) we can add a new cross-platform one while gradually deprecating the platform-specific one and not making any breaking changes.

File: src/FBProfile.js
Comment (leolusoli) [reply src/FBProfile.js:86]: I'm doing it right now.

**Comment by cocoa-rum** [comment]
@bernard-kms @leolusoli 
I think we are ready for review

_Also, what the best practice with package. I think, I am doing it a little bit complex way. It's my first time working with external package in such way
I added forked repo ( yarn add *local repo path*) Into my project. Do edits and paste changes into the package - commit_

File: ios/RCTFBSDK/login/RCTFBSDKLoginButtonManager.m
Comment (glenna) [inline ios/RCTFBSDK/login/RCTFBSDKLoginButtonManager.m:48]: style nit for consistency, can you put the opening bracket on a new line 🙏🏻

File: ios/RCTFBSDK/login/RCTFBSDKLoginButtonManager.m
Comment (glenna) [inline ios/RCTFBSDK/login/RCTFBSDKLoginButtonManager.m:52]: same here :)

File: ios/RCTFBSDK/login/RCTFBSDKLoginButtonManager.m
Comment (glenna) [inline ios/RCTFBSDK/login/RCTFBSDKLoginButtonManager.m:53]: Here, since we know `json` is a string, it is faster to use `-[NSString isEqualToString:]` method, so `[json isEqualToString:@"limited"]`

File: ios/RCTFBSDK/login/RCTFBSDKLoginManager.m
Comment (glenna) [inline ios/RCTFBSDK/login/RCTFBSDKLoginManager.m:79]: Same as above: `-[NSString isEqualToString:]`

File: src/FBProfile.js
Comment (glenna) [inline src/FBProfile.js:23]: I think this is a typo?

File: android/src/main/java/com/facebook/reactnative/androidsdk/Utility.java
Comment (thebergamo) [inline android/src/main/java/com/facebook/reactnative/androidsdk/Utility.java:155]: why do we need to place email here if it doesn't come from the endpoint itself?

File: android/src/main/java/com/facebook/reactnative/androidsdk/Utility.java
Comment (thebergamo) [inline android/src/main/java/com/facebook/reactnative/androidsdk/Utility.java:157]: I would just move this up as to have the data grouped in a logical sort, as you've the `firstName`, `middleName` and `lastName` on top

File: src/FBLoginManager.js
Comment (glenna) [inline src/FBLoginManager.js:74]: I see it is not standard in this lib, but other libs I work with also append the platform to these arguments to make it more clear that they are not used for all platforms. So similar as was done for the properties for FBLoginButton. But, as I'm not the maintainer here 🤷🏻‍♀️

File: ios/RCTFBSDK/core/RCTFBSDKProfile.m
Comment (thebergamo) [inline ios/RCTFBSDK/core/RCTFBSDKProfile.m:36]: On Android looks like we don't have this information coming from currentProfile, does iOS behave differently in such case? If yes, it would be nice to state it in the docs that only iOS should have value for this property.

File: ios/RCTFBSDK/core/RCTFBSDKProfile.m
Comment (thebergamo) [inline ios/RCTFBSDK/core/RCTFBSDKProfile.m:37]: Again, maybe moving this prop to the top of this return

File: types/FBProfile.d.ts
Comment (glenna) [inline types/FBProfile.d.ts:24]: According to the comments in the (Android) impl file, you actually cannot retrieve it via the profile, or?

**Review by thebergamo — COMMENTED** [review]
Besides my comments, there is also some warning about unused Platform variable!
Anyway, thanks very much for your contribution. Would like to see some comments from @bernard-kms if possible before integrate this with master ;)

**Review by glenna — COMMENTED** [review]
Hey @cocoa-rum, thanks for writing this feature! I'm definitely looking forward to this being released 😄 

I'm new to this repo/conversation, but thought I'd also provide a review since my team will be using this feature soon.

For me, it looks good overall, just some minor style nits, to keep consistency, and another naming convention that probably @thebergamo would like to weigh in on.

Also, just a small comment that on reviewing I did not understand why Profile and Authentication Token was added as part of the Limited Login functionality, until I went through and read the comments in the issue, so it would be good in the future to make that more apparently in the PR description as well.

Thank you adding this functionality! Looking forward to the release ;)

File: src/FBLoginManager.js
Comment (thebergamo) [reply src/FBLoginManager.js:74]: I totally agree with your point here :)

File: ios/RCTFBSDK/core/RCTFBSDKProfile.m
Comment (glenna) [reply ios/RCTFBSDK/core/RCTFBSDKProfile.m:36]: Good catch, I didn't catch this one

**Comment by cocoa-rum** [comment]
@thebergamo I got it. I will fix things that @glenna mentioned

> Also, just a small comment that on reviewing I did not understand why Profile and Authentication Token was added as part of the Limited Login functionality

As for Authentication Token - if we use limited login, we lose access to access token, so we need alternative and Authentication Token is. So from my view point Authentication Token complements limited log in

As for Profile, this [commnent](https://github.com/thebergamo/react-native-fbsdk-next/issues/7#issuecomment-819483813) 
And it appeared in this PR.
May be  we should ask @leolusoli  to create separate PR with Profile part? and from this PR we will remove it

File: src/FBLoginManager.js
Comment (cocoa-rum) [reply src/FBLoginManager.js:74]: What will be best practice for this?
Create **_logInWithPermissionsIOS_** and **_logInWithPermissionsAndroid_**

or something like this 
`logInWithPermissions ( { permissions,  loginTrackingIOS, nonceIOS } ) `

or did you mean
```
logInWithPermissions(
    permissions: Array<string>,
    loginTrackingIOS?: LoginTracking,
    nonceIOS?: string,
  )
```

Comment (glenna) [reply src/FBLoginManager.js:74]: @cocoa-rum

I mean the latter. So:

```
logInWithPermissions(
    permissions: Array<string>,
    loginTrackingIOS?: LoginTracking,
    nonceIOS?: string,
  )
```

**Comment by glenna** [comment]
@cocoa-rum thank you! yeah, I got the context when I read the issue, but as a reviewer I don't expect to need to read all comments on an issue thread to get the context ;). Just a future reference to add more information to the PR description.

Regarding moving Profile impl to a separate PR, I would leave that up to @thebergamo. (If it were me, I would leave it in, as now it's reviewed and ready to roll -- but update the description so that in the release notes for the version when the PR is linked there's easy to access information)

**Comment by thebergamo** [comment]
Regarding Profile impl, I believe it's pretty ok having it here, so no biggie from my side. You could just give more context as mentioned by @glenna in the PR description, so it would be much easier to track in the changelog later on :)

**Comment by leolusoli** [comment]
@glenna @thebergamo I'll explain you tomorrow with some examples of what happening for Profile **imageURL** field, because there's some particular situation regarding limited/not limited login (i.e. different value returned). I'll also fix the typos in comments and style some code according to the guidelines described in the comments. I think tomorrow will be all pushed and explained :)

Thanks for your time again.

File: android/src/main/java/com/facebook/reactnative/androidsdk/Utility.java
Comment (leolusoli) [reply android/src/main/java/com/facebook/reactnative/androidsdk/Utility.java:155]: Because the Profile class inside the Java SDK doesn't expose the email field. I place email=null for maintain the same behaviour with iOS. Do you think that is more correct to move this concept in JS?

File: ios/RCTFBSDK/core/RCTFBSDKProfile.m
Comment (leolusoli) [reply ios/RCTFBSDK/core/RCTFBSDKProfile.m:36]: I recognise now that even the iOS sdk retrieve the url with width and height (i.e. https://platform-lookaside.fbsbx.com/platform/profilepic/?asid=10224600543422918&height=100&width=100&....).
The iOS SDK set two default width and height values for the profile picture (100). In Android, we have only the method `getProfilePictureUri` with two params **mandatory**. So, I think that the correct way to handling this part is to pass 100 (as a default size setted by iOS) to the `getProfilePictureUri` in order to retrieve the same image for the two platforms. In the future (I think is out of scope of this PR), we can add a method that get the profile picture with the width and height parameters (even in iOS we have the possibility to make this). Could be the right way?

File: types/FBProfile.d.ts
Comment (leolusoli) [reply types/FBProfile.d.ts:24]: No, the email property is not inside the Profile implementation of the Android ([https://developers.facebook.com/docs/reference/androidsdk/current/facebook/com/facebook/profile.html/?locale=it_IT](url)). The only way to retrieve the email of the user is over GraphRequest. Maybe when Facebook will add the limited login in Android, they also add the email field in the Profile? Honestly, I don't know why this field is missing in Android.

Comment (bernard-kms) [reply types/FBProfile.d.ts:24]: I think `-IOS` suffix could be also applied to this one to inform about limitation in Android.

**Comment by cocoa-rum** [comment]
@thebergamo looks like we fixed all remarks for now
You are welcome to review or make some comments for improving PR

File: types/FBProfile.d.ts
Comment (leolusoli) [reply types/FBProfile.d.ts:25]: Sorry @bernard-kms, can you tell me where to place this suffix please?

Comment (bernard-kms) [reply types/FBProfile.d.ts:25]: I mean the `email` field, since this field is only valid for iOS as you mentioned.

Comment (leolusoli) [reply types/FBProfile.d.ts:25]: Mmmm... Honestly, i don't think this is a correct tip because, if you are don't asking the correct permissions (i.e. the "email" permission) for the method **loginInWithPermissions** you can't retrieve the value in the field even in iOS. I think that the values of the Profile are dependent to the type of the auth request (and his options) and not to the platform.

File: ios/RCTFBSDK/core/RCTFBSDKProfile.m
Comment (thebergamo) [reply ios/RCTFBSDK/core/RCTFBSDKProfile.m:37]: Instead of having default values I would just stick with them as mandatory and later on when we implement the method to retrieve state it clearly.
And yes, I believe it can be part of a new PR, maybe you could please just create an issue and link it here to not lose track of this `getProfilePictureUri` method as someone else could pick it :)

File: types/FBProfile.d.ts
Comment (thebergamo) [reply types/FBProfile.d.ts:25]: In that case we should have the value also on Android right? In such case I would stick with the suffix till we can maintain the same behavior in both platforms.

File: android/src/main/java/com/facebook/reactnative/androidsdk/Utility.java
Comment (thebergamo) [inline android/src/main/java/com/facebook/reactnative/androidsdk/Utility.java:104]: If you could, just in case would be nice to have some simple utility to putNull or value into the WritableMap to avoid repeating yourself that much in this case as you're having too many `if/else` similarly here.

**Review by thebergamo — APPROVED** [review]
From my perspective it's fine as it is, just minor changes that we can do after releasing this version. Although the email that comes only from iOS I would a more clear note in the `readme.md` about this feature and this property comes only from iOS for now as it's not supported by Android.
After having confirmation from others reviewers (cc @bernard-kms && @glenna) and feature documented I'm good to merge and release it :)

**Review by glenna — COMMENTED** [review]
Looking good!

I agree with @thebergamo about a section in the readme as well :)

File: types/FBLoginManager.d.ts
Comment (cocoa-rum) [reply types/FBLoginManager.d.ts]: Done. Even don't know, how I missed it

**Comment by glenna** [comment]
@thebergamo did you also want the readme section before merging? If so, I think that's the last thing we are waiting on.

**Comment by thebergamo** [comment]
Yes, because without that could lead others to some mistakes

**Comment by glenna** [comment]
@thebergamo agreed! Just wanted to make it clear :)

**Comment by cocoa-rum** [comment]
> Yes, because without that could lead others to some mistakes

Should I and @leolusoli  add changes to readme? Or what is next step?

**Comment by thebergamo** [comment]
Yes, you should.

**Comment by bernard-kms** [comment]
LGTM, and yes, a proper documentation is important to complete this work to be released.

**Comment by cocoa-rum** [comment]
Readme is ready. Is it correct?

**Review by glenna — APPROVED** [review]
looks good to me! Thank you! 🙏🏻

**Comment by thebergamo** [comment]
Version released :) [v4.2.0](https://github.com/thebergamo/react-native-fbsdk-next/releases/tag/v4.2.0)

Thank y'all for the effort for make it happen :)

**Pull request merged**
